### PR TITLE
[fix] fix scalar behavior for `torch.where`

### LINKED ISSUE #64161: The wrong behavior of ```torch.where```
## 🐛 The wrong behavior of ```torch.where``` function in PyTorch 1.9.0

When I utilize ```torch.where``` function to clip the tensor, I encounter an error as follows:
``` Python
RuntimeError: expected scalar type double but found float
```

## To Reproduce

The script is as follows:
```Python
import torch
print(torch.__version__)
x = torch.randn(3, 2)
print(x.dtype)
y = torch.where(x < 1e-8, 1e-8, x)  # <-- the error occurs
```

## Environment

PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.5.2 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.11)
CMake version: Could not collect
Libc version: N/A

Python version: 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46)  [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] (64-bit runtime)
Python platform: Darwin-20.6.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.9.0
[pip3] torchaudio==0.9.0a0+33b2469
[pip3] torchvision==0.10.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  h0a44026_0    pytorch
[conda] mkl                       2019.4                      233  
[conda] mkl-service               2.3.0            py36h9ed2024_0  
[conda] mkl_fft                   1.3.0            py36ha059aab_0  
[conda] mkl_random                1.0.4            py36h27c97d8_0  
[conda] numpy                     1.18.1                   pypi_0    pypi
[conda] numpy-base                1.17.0           py36h6575580_0  
[conda] pytorch                   1.9.0                   py3.6_0    pytorch
[conda] torchaudio                0.9.0                      py36    pytorch
[conda] torchvision               0.10.0                 py36_cpu    pytorch

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @nairbv @mruberry

### LINKED ISSUE #73298: torch.where does not handle scalar type correctly
### 🐛 Describe the bug

The following code:
```python
import torch

ab = torch.zeros(5, dtype=torch.bool)
af = torch.zeros(5, dtype=torch.float)
ad = torch.zeros(5, dtype=torch.double)

# This fail as expected
torch.where(ab, af, ad)  # RuntimeError: expected scalar type float but found double

# This works as expected
torch.where(ab, ad, 3.0)

# WTF are you doing!
torch.where(ab, af, 3.0)  # RuntimeError: expected scalar type float but found double
```
I am expecting `torch.where(ab, af, 3.0)` to work fine, because PyTorch only support double scalar, not float scalar. It does not make any sense to require the user to pass a float scalar as argument.

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0a0+gitec8d677
Is debug build: True
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: 13.0.1
CMake version: version 3.22.2
Libc version: glibc-2.35

Python version: 3.10.2 (main, Jan 15 2022, 19:56:27) [GCC 11.1.0] (64-bit runtime)
Python platform: Linux-5.16.9-arch1-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.55
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.54
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.1
/usr/lib/libcudnn_adv_infer.so.8.3.1
/usr/lib/libcudnn_adv_train.so.8.3.1
/usr/lib/libcudnn_cnn_infer.so.8.3.1
/usr/lib/libcudnn_cnn_train.so.8.3.1
/usr/lib/libcudnn_ops_infer.so.8.3.1
/usr/lib/libcudnn_ops_train.so.8.3.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] torch==1.12.0a0+gitec8d677
[pip3] torch-ucc==1.0.0
[pip3] torchani==2.2
[conda] Could not collect
```

cc @nairbv @mruberry

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #9190 
Fixes: #64161
Fixes https://github.com/pytorch/pytorch/issues/73298

Post this PR:
For scalar-tensor variant, the type-promotion rules apply and the dtype of output is determined after considering the dtypes of `x` and `y` of `torch.where` as per https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc
For scalar-scalar variant, we promote to the scalar with highest kind.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

### CODE REVIEW

**Comment by kshitij12345** [comment]
cc: @vadimkantorov

**Comment by vadimkantorov** [comment]
there are also similar type promotion issues with torch.lerp which is a "differentiable" version of torch.where. I think it wasn't prioritised, but I still hope it will be fixed as well :)

**Comment by kshitij12345** [comment]
Can you share the link to that issue :)
Thanks!

**Comment by vadimkantorov** [comment]
https://github.com/pytorch/pytorch/issues/57947

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

**Comment by kshitij12345** [comment]
Gentle ping @mruberry

**Comment by mruberry** [comment]
> Gentle ping @mruberry

Sorry I haven't had a chance to respond to this yet, @kshitij12345. I've asked @saketh-are to take a look as part of his study of type promotion. I'll be on pto starting Wednesday so you may need to be patient for a bit longer, though. Sorry about that.

**Comment by kshitij12345** [comment]
Hey Mike, thanks for the heads-up! No hurry. Enjoy the PTO :)

**Comment by vadimkantorov** [comment]
Does it also fix this?
```
>>> torch.where(torch.tensor([True]), torch.arange(1, dtype = torch.int16), 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected scalar type short int but found long int
```

**Comment by kshitij12345** [comment]
Yes
```python
>>> import torch
>>> torch.where(torch.tensor([True]), torch.arange(1, dtype = torch.int16), 1)
tensor([0], dtype=torch.int16)
>>> torch.where(torch.tensor([False]), torch.arange(1, dtype = torch.int16), 1)
tensor([1], dtype=torch.int16)
```

**Comment by mruberry** [comment]
Sorry for the delay, @kshitij12345, @saketh-are has been filling sick recently :(

**Comment by kshitij12345** [comment]
No hurry. Thanks for updating :)

File: test/test_torch.py
Comment (saketh-are) [inline test/test_torch.py:7632]: Would it be clearer if we reorganize these checks as follows?
```
if dtype.is_floating_point:
    return scalar_type in [int, float]
if dtype.is_complex:
    return True
return scalar_type == int
```

File: torch/_torch_docs.py
Comment (saketh-are) [inline torch/_torch_docs.py:10305]: It's great that we can get rid of this from the documentation now!

File: test/test_torch.py
Comment (saketh-are) [inline test/test_torch.py:7617]: Nice catch, if I understand correctly these loop variables were already unused?

**Comment by saketh-are** [comment]
Hey @kshitij12345, thanks a lot for your patience while I was out sick. This looks great and I left a few minor comments inline.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:7617]: Yes.

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py:7632]: Sure.

**Comment by kshitij12345** [comment]
Hey @saketh-are , 

Thanks for reviewing. Have simplified the valid combination logic. Should be ready for another round :)

File: test/test_torch.py
Comment (saketh-are) [inline test/test_torch.py:7665]: Super tiny nit: if/if would be better here (no elif) since the if path returns

**Review by saketh-are — APPROVED** [review]

File: test/test_torch.py
Comment (kshitij12345) [reply test/test_torch.py]: Done :)

**Comment by vadimkantorov** [comment]
Curious, does it also support 1) Python scalar 2) broadcasting as the condition argument? (Python scalar may be not very useful since `torch.where(cond, x, y)` can then be replaced by `x if cond else y`, but maybe useful for some tracing, not sure)

**Comment by kshitij12345** [comment]
> Curious, does it also support 1) Python scalar 2) broadcasting as the condition argument?

1. `condition` has to be a tensor. 
2. `condition` is broadcasted if necessary based on the shape of `x` or `y`.

**Comment by vadimkantorov** [comment]
Probably this PR fixes also https://github.com/pytorch/pytorch/issues/64161

**Comment by kshitij12345** [comment]
Gentle Ping @saketh-are @mruberry . This should be ready. Thanks!

**Comment by kshitij12345** [comment]
XLA failures are real. (I think it would be ok to skip XLA for now).

PTAL @mruberry @saketh-are

**Comment by JackCaoG** [comment]
@kshitij12345 Let me take a quick look at this, if this is something that xla can also we fix we should do that

**Comment by kshitij12345** [comment]
@JackCaoG Thanks for prompt look. Do note the the PR description is stale. The behaviour is different in latest commit. Will update the description tomorrow.

**Comment by JackCaoG** [comment]
@kshitij12345 I put my analysis in https://github.com/pytorch/xla/issues/3105#issuecomment-908574827. I think I know how to fix it, but I will wait until this pr being reviewed first.

**Comment by kshitij12345** [comment]
> @kshitij12345 I put my analysis in pytorch/xla#3105 (comment). I think I know how to fix it, but I will wait until this pr being reviewed first.

Your analysis looks correct. Will ping you once this PR is approved.

**Comment by kshitij12345** [comment]
@saketh-are have updated the PR so that the type promotion is supported for `tensor-scalar` variants. PTAL :)

File: aten/src/ATen/native/TensorCompare.cpp
Comment (saketh-are) [inline aten/src/ATen/native/TensorCompare.cpp:388]: This is interesting. Am I understanding correctly that it's due to https://github.com/pytorch/pytorch/pull/28427#issuecomment-545687342?

Comment (kshitij12345) [reply aten/src/ATen/native/TensorCompare.cpp:388]: That looks like the motivation. 
There is a check in `cpu_kernel` for the same (which fails without this casting here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/Loops.h#L257-L258
(called from here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L208

AFAIK, TensorIterator does similar things for CPU and type promotion

cc: @mruberry

**Comment by kshitij12345** [comment]
@saketh-are @mruberry JIT tracer failure looks real (so is XLA but @JackCaoG has taken care of it)

```
======================================================================
ERROR [3.530s]: test_where_ops (__main__.TestTEFuser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1642, in test_where_ops
    self.assertEqual(ref, t(cond, x, y))
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1596, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
  File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 2 element(s) (out of 2) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.0008750000000001812 (3.1415 vs. 3.140625), which occurred at index 0.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1646, in test_where_ops
    " ".join(["Failed:", str(dtype), op.__name__, device])
RuntimeError: Failed: torch.float16 <lambda> cuda
```

Looking at the test,
https://github.com/pytorch/pytorch/blob/8535418a06d75025541370cc656a8b6a0330ca0d/test/test_jit_fuser_te.py#L1616-L1624

It fails at first lambda (maybe fuser has different type promotion code for scalar case)

Should we skip the test for now or ping the fuser team?

Thanks!

File: aten/src/ATen/native/TensorCompare.cpp
Comment (saketh-are) [inline aten/src/ATen/native/TensorCompare.cpp:389]: If we're only promoting here on CPU, have you thought about what happens on CUDA? It seems that unpromoted self/other tensors would make it to the TensorIteratorConfig.

File: aten/src/ATen/native/TensorCompare.cpp
Comment (saketh-are) [inline aten/src/ATen/native/TensorCompare.cpp:384]: `at::result_type` is also implemented for (scalar, tensor) and (tensor, scalar) args. Would it make more sense to make use of that rather than using `wrapped_scalar_tensor_default_dtype` (which is a bit convoluted and also seems to have some comment saying not to use it)?

**Review by saketh-are — COMMENTED** [review]
Hey @kshitij12345, thanks for updating this PR to be more general! I left a couple of comments from discussion with @mruberry.

We were also wondering if you've thought about what happens in the out version if the provided `out` tensor has a "lesser" type than the chosen common dtype for the inputs.

Look forward to hearing your thoughts.

File: aten/src/ATen/native/TensorCompare.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/TensorCompare.cpp:389]: Even TensorIterator does not create temporary of `common_dtype` for GPU (since dynamic casting is supported on CUDA)

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/TensorIterator.cpp#L410-L449

cc: @mruberry (to correct if I am wrong in my understanding)

File: aten/src/ATen/native/TensorCompare.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/TensorCompare.cpp:384]: In that case we will need to add new overload for `where` kernel which takes a scalar as parameter (because the current version expects all inputs to be passed via TensorIterator which only takes Tensor inputs)  and add other overload for `_s_where` and corresponding entry in `derivatives.yaml`

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L204-L205

**Comment by kshitij12345** [comment]
> We were also wondering if you've thought about what happens in the out version if the provided out tensor has a "lesser" type than the chosen common dtype for the inputs.

I haven't considered that as there is no out version currently and I am not sure if we would add that in the future. The reason being that `out` shape for `where(Tensor condition) -> Tensor[]` overload is not static (and I feel it would be odd if `out` support is not consistent)

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/native_functions.yaml#L4542-L4562

Note: NumPy also doesn't have `out` support for ` where`.

**Comment by kshitij12345** [comment]
@mruberry 

Regarding our offline conversation about moving the type-promotion code to `where` instead of `_s_where`. I think that would be tricky. For eg. suppose `x` is `float` and `y` is `int64`. Our output should be float (which is allocated by `_s_where`. If these tensors reside on CUDA, then we don't need to cast these.
So for `_s_where` to know the `dtype` of output, it will have to check which input (passed by `where`) has higher dtype and allocate output accordingly (for which we will use `promote_types`. So we are sorta doing promotion computation again) 

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/aten/src/ATen/native/TensorCompare.cpp#L335

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/aten/src/ATen/native/TensorCompare.cpp#L359-L360

Plus I think it makes sense to have the type promotion logic closer to the kernel.

Let me know what you feel about the same (and if I missed something).

Thanks!

cc: @saketh-are

**Comment by mruberry** [comment]
> Regarding our offline conversation about moving the type-promotion code to `where` instead of `_s_where`. I think that would be tricky. For eg. suppose `x` is `float` and `y` is `int64`. Our output should be float (which is allocated by `_s_where`. If these tensors reside on CUDA, then we don't need to cast these.

You are correct. We could give up on the CUDA optimization or pass uncast tensors if the tensor is on CUDA (which we can test for in these functions).

> So for `_s_where` to know the `dtype` of output, it will have to check which input (passed by `where`) has higher dtype and allocate output accordingly (for which we will use `promote_types`. So we are sorta doing promotion computation again)

Yes without additional refactoring the computation has to be done again to determine the output's dtype (using result_type). That's imperfect but seems OK? Alternatively if the CUDA special case is dropped then the dtype of the output can be inferred from the dtype of any input tensor. 

> Plus I think it makes sense to have the type promotion logic closer to the kernel.

I agree but all these functions seem pretty close to the kernel.


What are your thoughts on giving up on fusing the CUDA copy or special-casing CUDA inputs?

**Comment by kshitij12345** [comment]
Have moved the promotion logic from `_s_where` to `where`. For now we have special cased on CUDA (to recompute the computation dtype for CUDA) in `_s_where`.

This actually took care of a bug related to type promotion. (Previously, the scalar tensor was broadcasted and hence took part in type promotion which should not be the case as per https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc)

Have also updated the scalar-scalar case to do promotion. This allows us to get rid of the hacky code.
https://github.com/pytorch/pytorch/blob/3324bae5f106f22ff78eb6943b79cd607a1b37e4/aten/src/ATen/native/TensorCompare.cpp#L230-L250

Thanks @mruberry @saketh-are!

This should be ready for another round.

**Comment by JackCaoG** [comment]
@kshitij12345 It seems like xla test pass, do I need to do anything on the pt/xla front?

**Comment by kshitij12345** [comment]
@JackCaoG the XLA tests are failing on the most recent run

https://app.circleci.com/pipelines/github/pytorch/pytorch/385123/workflows/cd4b647b-afdc-4901-8c75-ef529403f39e/jobs/16197732

**Comment by JackCaoG** [comment]
@kshitij12345 I see, I will take care of it.

File: aten/src/ATen/native/TensorCompare.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorCompare.cpp:311]: Add a comment explaining what this function is doing and how it's intended to be used and let's make the name specific to where ("get_promoted_where_tensor"?)

File: aten/src/ATen/native/TensorCompare.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorCompare.cpp:325]: Let's add a where type promotion note and reference it in places like this where the type promotion happens so readers can more easily understand what's going on

File: aten/src/ATen/native/TensorCompare.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorCompare.cpp:349]: This is going to need a comment (possibly a pointer to the where type promotion note referenced above).

Could this check be simplified to testing if self and other have distinct dtypes and, if they do, asserting they're not on the CPU?

File: aten/src/ATen/native/native_functions.yaml
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:4552]: Let's not extend this to be a method in this PR

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5464]: At least one of these cases should probably be scalar x tensor and scalar x scalar, right?

Can we include a complex scalar here, too?

File: aten/src/ATen/native/native_functions.yaml
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:4552]: Without this the newly added `cases_scalar` won't work for the method variant (for OpInfo tests). Also without this the function seems half baked.

```python
x.where(cond, y) # works
x.where(cond, 1) # doesn't work
```

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:5464]: Currently `input` has to be Tensor or a list of Tensor, thus we are checking only tensor x scalar case.

Other cases are tested in `test_torch`

File: aten/src/ATen/native/TensorCompare.cpp
Comment (mruberry) [inline aten/src/ATen/native/TensorCompare.cpp:326]: This PR should be sure to explain why scalar_tensor is called

https://github.com/pytorch/pytorch/blob/c0eb266c02c2886ada83662fad1107c8b5f23e1d/aten/src/ATen/native/TensorFactories.cpp#L618

as opposed to wrapped_scalar_tensor

https://github.com/pytorch/pytorch/blob/6c65ce8ee14faec577582ecd0e17ba436f32c8b3/aten/src/ATen/ScalarOps.h#L55

This assumes that it's OK to lose the information that the tensor is actually a wrapped number so long as where handles it own type promotion, right?

**Review by mruberry — COMMENTED** [review]
Overall looks really good to me, @kshitij12345. I made a few inline comments for your review. Looking forward to an update!

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5464]: Isn't the "input" tensor the boolean condition tensor, though?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:5464]: https://github.com/pytorch/pytorch/blob/4176afc4a035c0e79281b4955772c8962bc3ff4f/torch/testing/_internal/common_methods_invocations.py#L9651-L9655

File: aten/src/ATen/native/TensorCompare.cpp
Comment (JackCaoG) [reply aten/src/ATen/native/TensorCompare.cpp:311]: This will be helpful. I found it a bit hard to understand the promotion logic here. I would think we always promote both inputs to common_type but that doesn't seems to be what happens here.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5464]: Aha, tricky

**Comment by kshitij12345** [comment]
Gentle ping @saketh-are @mruberry

**Review by saketh-are — COMMENTED** [review]
Hey @kshitij12345, thanks for the updates on this non-trivial PR! I left a round of comments inline, mainly around ways we might be able to cut down on the complexity of the unit tests for `torch.where`. Look forward to seeing your thoughts on this.

**Comment by kshitij12345** [comment]
@saketh-are have mostly addressed all the reviews. I didn't exactly understand a few comments regarding test but have simplified the test reasonably (thanks for pointing that out :) ).

This should be ready for another round. Thanks!

**Review by saketh-are — COMMENTED** [review]
Hey @kshitij12345, thanks for making this round of changes. The new scalar-scalar tests look really clean, plus checking against `numpy.where` is always great for confidence in the correctness of our implementation. I'm wondering if we could extend the lists of scalars you made with some handmade tensors and cover the scalar-tensor case without much additional work there? Look forward to hearing your thoughts.

**Review by saketh-are — COMMENTED** [review]
Hey @kshitij12345, thanks for updating `test_where_scalar_scalar`! This is looking really good. Do you think we could eliminate `test_where_scalar` before we go ahead and merge this?

**Comment by saketh-are** [comment]
Thanks @kshitij12345 and @JackCaoG, please feel free to ping me once the XLA fix is ready.

**Comment by JackCaoG** [comment]
@saketh-are @kshitij12345 It seems like this fix also fix xla issue. I verified it with a small test and xla test seems to be passing on this pass. I think no change is required on the xla side.

**Comment by kshitij12345** [comment]
Thanks! @JackCaoG 

@saketh-are PTAL :)

**Comment by kshitij12345** [comment]
@JackCaoG I think this will need a fix at XLA side. 

The following reference test is failing : https://app.circleci.com/pipelines/github/pytorch/pytorch/403497/workflows/73e0bce0-bd07-4e77-860b-a5354ee5ff95/jobs/16643456?invite=true#step-107-3341

Also, the test from test_torch which used a fail previously is now merged into a test which is run only on Native Devices and hence get's skipped.

Depending on how complicated the fix is on XLA, we can either skip the tests on XLA for now or fix and unskip them. 

Thanks again!

**Comment by kshitij12345** [comment]
@JackCaoG gentle ping :)

**Comment by JackCaoG** [comment]
Hey @kshitij12345 Please check https://github.com/pytorch/pytorch/pull/62084#issuecomment-957064183. I think no change is required from the pt/xla side.

**Comment by JackCaoG** [comment]
@kshitij12345 Sorry, I didn't see your last comment. Let me check `test_reference_eager_where_xla_int64`

**Comment by JackCaoG** [comment]
FYI submitted https://github.com/pytorch/xla/pull/3208, waiting for review.

**Comment by JackCaoG** [comment]
@kshitij12345 PyTorch/XLA pr is ready, I can merge it once this pr is merged

**Comment by kshitij12345** [comment]
Thanks @JackCaoG 
@saketh-are XLA fix is ready. This PR should be ready for merge now. Thanks!

**Comment by vadimkantorov** [comment]
Hope ONNX export paths are sufficiently tested for these new overloads

**Comment by kshitij12345** [comment]
> Hope ONNX export paths are sufficiently tested for these new overloads

Thanks for pointing it out. I am not really sure of ONNX tests. Will have a look.

Any pointer will be helpful.

cc: @saketh-are @mruberry

**Comment by rgommers** [comment]
@saketh-are was there a problem landing this? @kshitij12345 can we give this another push and try to get it into `1.11.0`? There are multiple issues about this, and it also should address support for array API standard semantics.

**Comment by kshitij12345** [comment]
Gentle ping @saketh-are @mruberry

**Comment by JackCaoG** [comment]
@saketh-are Could we block this pr from merging for maybe 1 day? It has been too long and downstream xla pr has a bunch of merge conflicts and I am fixing that. I will circle back when I fix the conflict and verify the fix still work.

**Comment by vadimkantorov** [comment]
Cool that this is getting merged! Would also be nice to have similar type promotion for `torch.lerp` (convenient for blending images): https://github.com/pytorch/pytorch/issues/57947, https://github.com/pytorch/pytorch/issues/54389

**Comment by JackCaoG** [comment]
@saketh-are XLA pr is ready, feel free to merge this one.

**Comment by kshitij12345** [comment]
@saketh-are Gentle ping :)

**Comment by kshitij12345** [comment]
@saketh-are ping :)

**Comment by kshitij12345** [comment]
@saketh-are @mruberry Ping :) Let me know if I should resolve the conflicts

**Comment by vadimkantorov** [comment]
@ngimel Is this superseded by https://github.com/pytorch/pytorch/pull/76691 ?

**Comment by ngimel** [comment]
yes

**Comment by vadimkantorov** [comment]
just hope all the test cases from #9190 are added

**Pull request closed without merging**
